Userland/Applications/Piano/WaveWidget.cpp

```diff
@@ -35,19 +51,20 @@ void WaveWidget::paint_event(GUI::PaintEvent& event)
 
     Color left_wave_color = left_wave_colors[m_track_manager.current_track()->synth()->wave()];
     Color right_wave_color = right_wave_colors[m_track_manager.current_track()->synth()->wave()];
-    // FIXME: We can't get the last buffer from the track manager anymore
-    auto buffer = FixedArray<Music::Sample>::must_create_but_fixme_should_propagate_errors(sample_count);
+    auto buffer = FixedArray<Audio::Sample>::must_create_but_fixme_should_propagate_errors(sample_count);
+    m_track_manager.current_track()->current_signal(buffer);
```

Comment:
First of all, you're allocating a new buffer every frame, leading to a lot of heap churn. If you look around at how audio buffers are normally used, they're only reallocated in special situations.

And second, current_signal runs the DSP processing pipeline for the entire track, which may be an extremely expensive operation that is offloaded to a separate thread for a reason. A better idea would be to grab the last buffer from the track manager and store it there if necessary; I'll make sure to keep such an API around when TrackManager moves to LibDSP.

Reply:
Okay I need to look into that further as I'm unsure what grabbing the last buffer means.

On the other hand, isn't it already the case that it'll cause heap churn as it already creates a `FixedArray` but doesn't do anything with it?

Reply:
> Okay I need to look into that further as I'm unsure what grabbing the last buffer means.

Whenever TrackManager renders an audio chunk, it could store the result until the next chunk is rendered. This would only incur one allocation at the beginning, as you could overwrite the previous saved result with the new one.

> On the other hand, isn't it already the case that it'll cause heap churn as it already creates a `FixedArray` but doesn't do anything with it?

I'm not sure where we create an unused FixedArray here or in TrackManager.